wgpu-hal/src/gles/wgl.rs

```diff
@@ -0,0 +1,748 @@
+                String::from("unable to get the device context from window"),
+                Error::last_os_error(),
+            ));
+        }
+        Ok(Surface {
+            dc,
+            presentable: true,
+            swapchain: None,
+            srgb_capable: self.srgb_capable,
+        })
+    }
+    unsafe fn destroy_surface(&self, _surface: Surface) {}
+
+    unsafe fn enumerate_adapters(&self) -> Vec<crate::ExposedAdapter<super::Api>> {
+        unsafe {
+            super::Adapter::expose(AdapterContext {
+                inner: self.inner.clone(),
+            })
+        }
+        .into_iter()
+        .collect()
+    }
+}
+
+pub struct Swapchain {
+    surface_context: WglContext,
+    surface_gl: glow::Context,
+    framebuffer: glow::Framebuffer,
+    renderbuffer: glow::Renderbuffer,
+    /// Extent because the window lies
+    extent: wgt::Extent3d,
+    format: wgt::TextureFormat,
+    format_desc: super::TextureFormatDesc,
+    #[allow(unused)]
+    sample_type: wgt::TextureSampleType,
+}
+
+pub struct Surface {
+    // FIXME: Find a way to call ReleaseDC on the same thread as GetDC.
+    dc: HDC,
+    pub(super) presentable: bool,
+    swapchain: Option<Swapchain>,
+    srgb_capable: bool,
+}
+
+unsafe impl Send for Surface {}
+unsafe impl Sync for Surface {}
+
+impl Surface {
+    pub(super) unsafe fn present(
+        &mut self,
+        _suf_texture: super::Texture,
+        context: &AdapterContext,
+    ) -> Result<(), crate::SurfaceError> {
+        let sc = self.swapchain.as_ref().unwrap();
+        // Hold the lock for the shared context as we're using resources from there.
+        let _inner = context.inner.lock();
+
+        if let Err(e) = sc.surface_context.make_current() {
+            log::error!("unable to make the surface OpenGL context current: {e}",);
+            return Err(crate::SurfaceError::Other(
+                "unable to make the surface OpenGL context current",
+            ));
+        }
+
+        let gl = &sc.surface_gl;
+
+        // Note the Y-flipping here. GL's presentation is not flipped,
+        // but main rendering is. Therefore, we Y-flip the output positions
+        // in the shader, and also this blit.
+        unsafe {
+            gl.blit_framebuffer(
+                0,
+                sc.extent.height as i32,
+                sc.extent.width as i32,
+                0,
+                0,
+                0,
+                sc.extent.width as i32,
+                sc.extent.height as i32,
+                glow::COLOR_BUFFER_BIT,
+                glow::NEAREST,
+            )
+        };
+
+        if unsafe { SwapBuffers(self.dc) } == FALSE {
+            log::error!("unable to swap buffers: {}", Error::last_os_error());
+        }
+
+        Ok(())
+    }
+
+    pub fn supports_srgb(&self) -> bool {
+        self.srgb_capable
+    }
+}
+
+impl crate::Surface<super::Api> for Surface {
+    unsafe fn configure(
```

Comment:
Missed a thing here:

This needs to be documented in the wgpu-hal api, but `configure` is expected to clean up after the previous configuration if there was one. `unconfigure` is only called when the surface is destroyed.